app/utils/callback/lava.py

```diff
@@ -621,106 +627,52 @@ def add_tests(job_data, job_meta, lab_name, db_options,
-                job_tc = _get_test_case(suite_results, ('job',))
-                login_line_num = _get_log_line_number(log,
-                                                      LOGIN_CASE_END_PATTERN)
-                start_log_line = 0 if login_line_num is None \
-                    else login_line_num
-            else:
-                suite_name = suite_name.partition("_")[2]
-                group = dict(meta)
-                group[models.NAME_KEY] = suite_name
-                _add_test_results(group, suite_results,
-                                  end_lines_map)
-                groups.append(group)
-        if login_tc and login_tc.get('result') == 'pass' and len(groups) == 0:
-            login_tc = job_tc
-        if login_tc:
-            _add_login_case(meta, cases, login_tc)
-        add_log_fragments(groups, log, end_lines_map, start_log_line)
-
-        if (len(groups) == 1) and (groups[0][models.NAME_KEY] == plan_name):
-            # Only one group with same name as test plan
-            plan = groups[0]
-            if cases:
-                insert_len = len(cases)
-                plan_cases = plan[models.TEST_CASES_KEY]
-                cases.extend(plan_cases)
-                plan[models.TEST_CASES_KEY] = cases
-        elif groups or cases:
-            # Create top-level group with the test plan name
-            plan = dict(meta)
-            plan[models.NAME_KEY] = plan_name
-            plan[models.SUB_GROUPS_KEY] = groups
-            plan[models.TEST_CASES_KEY] = cases
-
-        if plan:
-            ret_code, plan_doc_id, err = \
-                utils.kci_test.import_and_save_kci_tests(plan, db_options)
-            utils.errors.update_errors(errors, err)
-    except (yaml.YAMLError, ValueError) as ex:
-        ret_code = 400
-        msg = "Invalid test data from LAVA callback"
+        store_artifacts(callback.meta, job_data, callback.log)
     except (OSError, IOError) as ex:
         ret_code = 500
         msg = "Internal error"
-    finally:
-        if ex is not None:
-            utils.LOG.exception(ex)
-        if msg is not None:
-            utils.LOG.error(msg)
-            utils.errors.add_error(errors, ret_code, msg)
-        if errors:
-            raise utils.errors.BackendError(errors)
+        utils.errors.add_error(errors, ret_code, msg)
+        handle_errors(ex, msg, errors)
+
+    test_results = LavaResults(callback.results,
+                               callback.meta,
+                               callback.log)
+    plan = LavaPlan(test_results.groups, test_results.cases, callback.meta)
+
+    if plan.data:
+        ret_code, plan_doc_id, err = \
+            utils.kci_test.import_and_save_kci_tests(plan.data, db_options)
+        utils.errors.update_errors(errors, err)
+        handle_errors(errors=errors)
 
     if not plan_doc_id:
         utils.LOG.warn("No test results")
         return None
 
     return plan_doc_id
+
+
+def handle_errors(ex=None, msg=None, errors=None):
```

Comment:
Maybe this should be in `utils.errors`?

Reply:
right, it fits better in utils.errors. I moved it.